numpy/core/src/umath/ufunc_object.c

```diff
@@ -3531,8 +3534,50 @@ reduce_loop(NpyIter *iter, char **dataptrs, npy_intp *strides,
         strides_copy[0] = strides[0];
         strides_copy[1] = strides[1];
         strides_copy[2] = strides[0];
-        innerloop(dataptrs_copy, countptr,
-                    strides_copy, innerloopdata);
+
+        if (!masked) {
+            innerloop(dataptrs_copy, countptr,
+                      strides_copy, innerloopdata);
+        }
+        else {
+            npy_intp count = *countptr;
+            char *maskptr = dataptrs[2];
+            npy_intp mask_stride = strides[2];
+            char mask = *maskptr;
+            if (mask_stride != 0) {
+                /*
+                 * Process the chunks of the data where the mask is set;
+                 * Here, we optimize for the hopefully common case where
+                 * most elements are included, i.e., there are large chunks
+                 * of data to be looped over.
+                 */
+                npy_intp i;
+                npy_intp iprev = 0;
+
+                for (i = 0; i < count; i++, maskptr += mask_stride) {
+                    /* check for a change in mask */
+                    if (*maskptr != mask) {
+                        npy_intp n = i - iprev;
+                        if (mask) {
+                            innerloop(dataptrs_copy, &n,
+                                      strides_copy, innerloopdata);
+                        }
+                        dataptrs_copy[0] += n * strides[0];
+                        dataptrs_copy[1] += n * strides[1];
+                        dataptrs_copy[2] = dataptrs_copy[0];
+                        mask = *maskptr;
+                        iprev = i;
+                    }
+                }
+                /* The number of items not yet processed */
+                count -= iprev;
+            }
+            /* If either the broadcast mask or the last one is false, we're done */
+            if (mask) {
+                innerloop(dataptrs_copy, &count,
+                          strides_copy, innerloopdata);
+            }
```

Comment:
Possibly clearer to transpose the two loops here:
```C++
npy_intp i = 0;
while (i < count) {
    /* Find where the mask changes */
    char mask = *maskptr;
    npy_intp i_next;
    if (mask_stride == 0) {
        /* optimization */
        i_next = count;
    }
    else {
        i_next = i;
        while (mask == *maskptr && i_next < count) {
            i_next++;
            maskptr += mask_stride;
        }
    }
    /* Copy the last contiguous region */
    npy_intp n = i_next - i;
    if (mask) {
        innerloop(dataptrs_copy, &n,
                  strides_copy, innerloopdata);
    }
    dataptrs_copy[0] += n * strides[0];
    dataptrs_copy[1] += n * strides[1];
    dataptrs_copy[2] = dataptrs_copy[0];
    i = i_next;
}
```

You can also leave the stride == 0 optimization to the compiler:
```C
while (count > 0) {
    /* Find where the mask changes */
    char mask = *maskptr;
    npy_intp n = find(maskptr, count, mask_stride, !mask);

    /* Copy the last contiguous region */
    if (mask) {
        innerloop(dataptrs_copy, &n,
                  strides_copy, innerloopdata);
    }
    dataptrs_copy[0] += n * strides[0];
    dataptrs_copy[1] += n * strides[1];
    maskptr          += n * mask_stride
    dataptrs_copy[2] = dataptrs_copy[0];
    count -= n;
}

static inline npy_intp
find(npy_bool *data, npy_intp n, npy_intp stride, bool value) {
    /* let the compiler optimize for zero stride */
    if (stride == 0){
        return find_inner(data, n, stride, value);
    }
    else {
        return find_inner(data, n, stride, value);
    }
}

static inline npy_intp
find_inner(npy_bool *data, npy_intp n, npy_intp stride, bool value) {
    for (npy_intp i = 0; i < n; i++, data += stride) {
        if (*data == value) {
            return i;
        }
    }
    return n;
}
```

Reply:
I like the version that does not rely on the compiler to optimize and adopted an even simpler version of it. It makes the case where the mask is all-true quite a bit faster:
```
a = np.arange(100000.)
m = ~np.isnan(a)
%timeit np.add.reduce(a, where=m)
# 90 -> 70 us
```
Adding either `a[::100] = np.nan` or `a[::2] = np.nan`, the timings stay basically the same.

Reply:
To be clear - changing from the version this comment was added on to the latest version gives that speed boost? Or the `stride=0` optimization gives that speed up in either case?

Reply:
The change in version (i.e., the last commit) is what gave the speed boost - it has nothing to do with the `stride=0` optimization (the mask has a stride in my example, it is just set to all True). I'm actually rather confused why there is such a speed-up, it must be that the compiler can make the inner `where` loop substantially faster than the large loop with a complex but unlikely if clause.

p.s. For the record, the timings for the `stride=0` case do not change.

Reply:
> I'm actually rather confused why there is such a speed-up

So was I, hence the request for clarification!